proposals/xxxx-wave-matrix.md

```diff
@@ -0,0 +1,396 @@
+
+### WaveMatrix Matrix Objects
+
+The code below approximately declares the base interface that WaveMatrix matrix
+objects implement.
+
+```c++
+namespace detail {
+template <typename ElTy, int NRows, int NCols>
+class WaveMatrixBase {
+  void Fill(ElTy Val);
+  void Load(ByteAddressBuffer Res, uint StartOffset, uint Stride, bool ColMajor,
+            uint Align = 0);
+  void Load(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+            bool ColMajor, uint Align = sizeof(ElTy));
+
+  void Load(groupshared ElTy Arr[], uint StartIdx, uint Stride, bool ColMajor);
+
+  void Store(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+             bool ColMajor, uint Align = sizeof(ElTy));
+
+  void Store(groupshared ElTy Arr[], uint StartIdx, uint Stride, bool ColMajor);
+};
+} // namespace detail
+```
+
+#### Loading and Storing WaveMatrix Matrix Objects
+
+Values for WaveMatrix matrix objects can be loaded from and stored to
+`[RW]ByteAddressBuffer` or `groupshared` array of type `ElTy`.
+
+Loading from or storing to a `[RW]ByteAddressBuffer` takes a start offset in
+bytes, row stride in bytes, and optional row alignment in bytes. Rows begin at
+algined offsets from the `StartOffset` based on the `Stride` and `Alignment`
+(`StartOffset + (RowIdx * [(Stride % Alignment) + Stride])`).
+
+The `Alignment` must be at least `sizeof(ElTy)` otherwise the load behavior is
+undefined.
+
+Loading from or storing to a `groupshared` array takes the starting index, and
+the row stride as a number of elements.
+
+##### Orientation
+
+When loading and storing WaveMatrix matrices a boolean parameter is provided to
+indicate if the matrix being loaded or stored is column major.
+
+Matrices may be stored in row or column layout. Matrices are always loaded into
+row-major orientation in memory for WaveMatrix objects. The `Load` and `Store`
+methods perform matrix transposition when loading or storing column major
+matrices.
+
+##### Stride
+
+When loading and storing from `groupshared` arrays, the stride is expressed in
+number of elements.
+
+When loading and storing from `[RW]ByteAddressBuffer` types, the stride is
+expressed in bytes. The row stride must be a multiple of the size of the
+element, and greater than or equal to the size of the element multiplied by the
+number of elements per row. Any value below the minimum legal values are
+ignored. The behavior of row stride values that are not a multiple of element
+stride is undefined.
+
+#### WaveMatrix Left & Right
+
+```c++
+template <typename ElTy, int NRows, int NCols>
+class WaveMatrixLeft : detail::WaveMatrixBase<ElTy, NRows, NCols> {
+  uint MatrixDepth();
+};
+
+template <typename ElTy, int NRows, int NCols>
+class WaveMatrixRight : detail::WaveMatrixBase<ElTy, NRows, NCols> {
+  uint MatrixDepth();
+};
+```
+
+`ElTy` must be either a 32 or 16 bit floating point type or an 8 bit packed
+signed or unsigned integer type. `NRows` and `NCols` must be compile-time
+constant expressions.

```

Comment:
It is fairly obvious, but perhaps stating that NRows is the number of rows and NCols the number of columns.

I think we should explain what Left and Right matrices are and how they can be used

Reply:
I want to be a little careful about how deep this document goes in terms of usage. This isn't intended to be user documentation, this is a specification of the language feature.